src/Components/Components/src/Routing/Router.cs

```diff
@@ -93,10 +101,21 @@ public Task SetParametersAsync(ParameterView parameters)
                 throw new InvalidOperationException($"The {nameof(Router)} component requires a value for the parameter {nameof(NotFound)}.");
             }
 
-
             var assemblies = AdditionalAssemblies == null ? new[] { AppAssembly } : new[] { AppAssembly }.Concat(AdditionalAssemblies);
             Routes = RouteTableFactory.Create(assemblies);
+
+            // If we're about to render the router for the first time, then
+            // we need to call the `OnNavigate` handler to ensure that pre-processing
+            // steps are completed before rendering the route. This way, it will work
+            // if you navigate to /PageWithLazyLoadedAssemblies or visit it for the first time.
+            if (OnNavigate != null && !initialOnNavigateCalled)
+            {
+                OnNavigate(NavigationManager.ToBaseRelativePath(_locationAbsolute));
+                initialOnNavigateCalled = true;
+            }
```

Comment:
What's different between this and https://github.com/dotnet/aspnetcore/pull/23290/files#diff-c9e6808455b40dcab1c29363164978afR182-R189

Could you not do here something like 
```csharp
if(OnNavigate != null && !OnNavigate(NavigationManager.ToBaseRelativePath(_locationAbsolute))
{
  return;
}
```

I think you don't need to track the initialization, do you? If you navigate directly to a "lazy route" then it will do the right thing? (Stop, trigger the load and then once it has loaded it will refresh the router and continue?)

Reply:
`SetParametersAsync` gets called after the user invokes `StateHasChanged`. If there isn't a guard to ensure that the `OnNavigate` handler is called at most once on each render then it'll keep being called.

Reply:
> `SetParametersAsync` gets called after the user invokes `StateHasChanged`. If there isn't a guard to ensure that the `OnNavigate` handler is called at most once on each render then it'll keep being called.

I'm not sure I understand, I expect this to return false the first time it calls OnNavigate have the callback trigger whatever action it needs to load the assemblies, trigger the state change after the assemblies are loaded (which sets the new loaded assemblies as parameters) and return true the next time it gets called with the same path since the assemblies have already been loaded.

Am I missing something?